### PR TITLE
ci: use `bzlmod` in most Bazel builds

### PR DESCRIPTION
**Pull request by coryan**
Part of the work for #11485

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14401)

### CODE REVIEW

File: MODULE.bazel [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+bazel_dep(name = "platforms", version = "0.0.10")
+bazel_dep(name = "bazel_skylib", version = "1.7.1")
+bazel_dep(name = "rules_cc", version = "0.0.9")
+bazel_dep(name = "abseil-cpp", version = "20240116.2", repo_name = "com_google_absl")
+bazel_dep(name = "protobuf", version = "27.2", repo_name = "com_google_protobuf")
+bazel_dep(name = "boringssl", version = "0.0.0-20230215-5c22014")
+bazel_dep(name = "grpc", version = "1.63.1", repo_name = "com_github_grpc_grpc")
+bazel_dep(name = "nlohmann_json", version = "3.11.3", repo_name = "com_github_nlohmann_json")
+bazel_dep(name = "crc32c", version = "1.1.0", repo_name = "com_github_google_crc32c")
+bazel_dep(name = "opentelemetry-cpp", version = "1.14.2", repo_name = "io_opentelemetry_cpp")
```
Comment (dbolduc) [inline MODULE.bazel:35]: FYI to both of us: I think this will conflict with https://github.com/googleapis/google-cloud-cpp/pull/14416

Comment (coryan) [reply MODULE.bazel:35]: We could (should?) send a PR to upgrade `opentelemetry-cpp`, then we don't need to worry about the flag anymore.

**Comment by coryan** [comment]
PTAL

**Review by dbolduc — APPROVED** [review]

**Pull request merged**